File: JavaScript/79-Word Search.js
var exist = function(board, word) {
    let row = board.length, col = board[0].length;
    
    let dfs = (word, r, c) =>{
        if (!word.length){
            return true;
        }
        else if (r>=0 && r<row && c>=0 && c<col && word[0] == board[r][c]){

            let tmp = board[r][c];
            board[r][c] = '#';
            if (dfs(word.substr(1), r+1, c) || dfs(word.substr(1), r-1, c) || dfs(word.substr(1), r, c+1) || dfs(word.substr(1), r, c-1)){
                return true;
            }
            board[r][c] = tmp;
            return false;
        }
        else {
            return false;
        }
    }
    for (let r=0; r<row; r++){
        for (let c=0; c<col; c++){
            if (dfs(word, r, c)){
                return true;
            }
        }
    }
    return false;
};

File: C++/79-Word Search.cpp
// O(row*col*3^L)
class Solution {
public:
    Solution() {
        dirs = {{0, -1}, {-1, 0}, {0, 1}, {1, 0}};
    }
    int N, row, col;
    bool visited[10][10];
    bool dfs(int r, int c, int wordIdx, vector<vector<char>>& board, string& word) {
        if (word[wordIdx] != board[r][c]) return false;
        if (word[wordIdx] == board[r][c] && wordIdx == N-1) return true;
        visited[r][c] = true;
        for (const auto& [dr, dc]:dirs) {
            int nr = r+dr, nc = c+dc;
            if (0<=nr && nr<row && 0<=nc && nc<col && !visited[nr][nc]) {
                if (dfs(r+dr, c+dc, wordIdx+1, board, word)) {
                    return true;
                }
            }
        }
        visited[r][c] = false;
        return false;
    }
    
    bool exist(vector<vector<char>>& board, string word) {
        this->row = board.size();
        this->col = board[0].size();
        this->N = word.size();

        for (int r=0; r<row; r++) {
            for (int c=0; c<col; c++) {
                if (dfs(r, c, 0, board, word)) return true;
            }
        }
        
        return false;
    }
private:
    vector<pair<int, int>> dirs;
};